File: Code/Python/fundamentals/BankAccount.py
class BankAccount:
    accounts = []

    def __init__(self, int_rate, balance=0):
        self.interest_rate = int_rate
        self.balance = balance
        BankAccount.accounts.append(self)

    def deposit(self, amount):
        self.balance += amount
        return self

    def withdraw(self, amount):
        if self.balance >= amount:
            self.balance -= amount
        else:
            print("Insufficient funds: Charging a $5 fee")
            self.balance -= 5
        return self

    def display_account_info(self):
        print("Balance: $" + str(self.balance))
        return self

    def yield_interest(self):
        if self.balance > 0:
            self.balance += self.balance * self.interest_rate
        return self

    @classmethod
    def print_all_accounts_info(cls):
        for account in cls.accounts:
            account.display_account_info()

account1 = BankAccount(0.01, 1000)
account2 = BankAccount(0.01, 500)

account1.deposit(100).deposit(200).deposit(300).withdraw(150).yield_interest().display_account_info()

account2.deposit(50).deposit(100).withdraw(20).withdraw(30).withdraw(40).withdraw(50).yield_interest().display_account_info()

BankAccount.print_all_accounts_info()

File: Code/Java/OOP/BankAccount/BankAccount.java

public class BankAccount {
    // MEMBER VARIABLES
    private double checkingBalance;
    private double savingsBalance;
    private String accountNumber;

    private static int accounts;
    private static double totalMoney; // refers to the sum of all bank account checking and savings balances

    // CONSTRUCTOR
    // Be sure to increment the number of accounts
    public BankAccount() {
        this.checkingBalance = 0.0;
        this.savingsBalance = 0.0;
        accounts++;
    }

    // GETTERS
    // for checking, savings, accounts, and totalMoney
    public double getCheckingBalance() {
        return this.checkingBalance;
    }

    public double getSavingsBalance() {
        return this.savingsBalance;
    }

    public static int getNumberOfAccounts() {
        return accounts;
    }

    public static double getTotalMoney() {
        return totalMoney;
    }

    public String getAccountNumber() {
        return this.accountNumber;
    }

    // METHODS
    // deposit
    // - users should be able to deposit money into their checking or savings
    // account
    public void deposit(String accountType, double amount) {
        if (amount > 0) {
            if (accountType.equals("checking")) {
                checkingBalance += amount;
            } else if (accountType.equals("ssaving")) {
                savingsBalance += amount;
            }
            totalMoney += amount;
            System.out.println("Deposit $" + amount + " into " + accountType + " account.");
        } else {
            System.out.println("Invalid deposit amount.");
        }
    }

    // withdraw
    // - users should be able to withdraw money from their checking or savings
    // account
    // - do not allow them to withdraw money if there are insufficient funds
    // - all deposits and withdrawals should affect totalMoney
    public void withdraw(String accountType, double amount) {
        if (amount > 0) {
            if (accountType.equals("checking")) {
                if (checkingBalance >= amount) {
                    checkingBalance -= amount;
                    totalMoney -= amount;
                    System.out.println("Withdrawn $" + amount + " from " + accountType + " account.");
                } else {
                    System.out.println("Insufficient funds in checking account.");
                }
            } else if (accountType.equals("savings")) {
                if (savingsBalance >= amount) {
                    savingsBalance -= amount;
                    totalMoney -= amount;
                    System.out.println("Withdrawn $" + amount + " from " + accountType + " account.");
                } else {
                    System.out.println("Insufficient funds in savings account.");
                }
            } else {
                System.out.println("Invalid account type.");
            }
        } else {
            System.out.println("Invalid withdrawn amount.");
        }
    }
    // getBalance
    // - display total balance for checking and savings of a particular bank account
    public void getBalance() {
        System.out.println("Checking Account Balance: $" + checkingBalance);
        System.out.println("Savings Account Balance: $" + savingsBalance);
    }
}
